### PR TITLE
Bolt re-runs

### PR DESCRIPTION
**Pull request by jgoizueta**
Our bolt lock, which is used by Ghost Tables Manager (GTM) and also by `UserTableIndexService`, has a retry mode (introduced in 3d8801c) with the intention of not missing schema changes that should be handled by GTM.

I think this is being handled incorrectly:
* We're using it when locking the GTM to synchronize syncs/imports: in this cases there's no need to rerun GTM
* When the lock is not acquired we signal a re-run is pending by setting a redis key (`*:retry`). so that the running GTM runs again when it finishes. The problem with that is that the GTM process could keep running indefinitely if retries keep being flagged regularly, holding a resque job.

The changes here change the retry behaviour: when the lock cannot be acquired to run GTM a new run is enqueued unless there's one already enqueued and not executed. We use the `*:retry` flags now to signal that we have enqueued the job and it hasn't been executed. So this acts as a retry-debounce mechanism for GTM calls and performs reruns asynchronously at a later time given a chance to other resque/GTM jobs to run.

There's another related issue that I'm not fixing here: We try to get the GTM lock in imports and syncs but run them any way even if the lock is not acquired. I think this is unnecessary but I'm leaving the code in place because it emits a rollback trace when that happens that can be interesting in debugging some cases and in watching the effects of putting this in production (as it shows the frequency of concurrency of syncs with GTM for some users with many syncs).

Note that `UserTableIndexService` doesn't use the bolt retries, so these changes only affect GTM.

[EDIT: I didn't understand initially how rerun was doing its job, I've updated the comments here]

### CODE REVIEW

**Comment by jgoizueta** [comment]
Tests need to be adjusted. Working on it

File: lib/carto/bolt.rb [hunk 1]
```diff
@@ -20,13 +20,16 @@ def run_locked(attempts: DEFAULT_RETRY_ATTEMPTS,
...
       begin
-        unless locked_acquired
-          set_rerun_after_finish
-          return !!locked_acquired
+        if locked_acquired
+          remove_retry
+          yield
+          true
+        else
+          if rerun_func && !set_retry
```
Comment (gonzaloriestra) [inline lib/carto/bolt.rb]: `rerun_func` is not a function to run after the main task anymore, but a function to run when it's not possible to get the lock. So I think we should rename it, although I don't have a good alternative... :sweat_smile: What about `fail_function`?

File: lib/carto/ghost_tables_manager.rb [hunk 2]
```diff
@@ -49,8 +53,7 @@ def get_bolt
     def self.run_synchronized(user_id, attempts: 10, timeout: 30000, **warning_params)
       gtm = new(user_id)
       bolt = gtm.get_bolt
-      rerun_func = lambda { gtm.send(:sync) }
```
Comment (gonzaloriestra) [inline lib/carto/ghost_tables_manager.rb]: With your current changes, I agree that a sync job should be always enqueued at some point and run later when the lock is released. But I think the idea of this approach was to sync right after it finishes, without any delay. Anyway, this is used for long processes like import, so I don't think it's a problem to enqueue it and wait just a bit more.

File: lib/carto/bolt.rb [hunk 3]
```diff
@@ -75,16 +70,12 @@ def get_lock
...
     end
 
-    def set_rerun_after_finish
-      @redis_object.set("#{@bolt_key}:retry", true, px: @ttl_ms, nx: true)
-    end
-
-    def refresh_lock_timeout
-      @redis_object.pexpire(@bolt_key, @ttl_ms)
+    def remove_retry
+      @redis_object.del("#{@bolt_key}:retry")
```
Comment (gonzaloriestra) [inline lib/carto/bolt.rb]: I think `retried` would be more accurate, what do you think?

File: lib/carto/bolt.rb [hunk 4]
```diff
@@ -20,13 +20,16 @@ def run_locked(attempts: DEFAULT_RETRY_ATTEMPTS,
       locked_acquired = acquire_lock(attempts, timeout)
 
       begin
-        unless locked_acquired
-          set_rerun_after_finish
-          return !!locked_acquired
+        if locked_acquired
```
[the remark this answers is not in the document]
Comment (jgoizueta) [reply lib/carto/bolt.rb]: I think that's not allowed for variables (the `?` suffix), only for methods

[the remark this answers is not in the document]
Comment (gonzaloriestra) [reply lib/carto/bolt.rb]: I meant to rename it to `lock_acquired` (the question mark is outside the code block :slightly_smiling_face:)

[the remark this answers is not in the document]
Comment (jgoizueta) [reply lib/carto/bolt.rb]: OMG I hadn't notice the difference `locked` vs `lock`!!

**Comment by jgoizueta** [comment]
With this change we're now using the `:retry` redis key as a key to indicate there's a GTM job queued and pending. We could rename it to something like `:pending` for clarity and use it systematically to avoid enqueing multiple jobs per user, i.e.

Here: https://github.com/CartoDB/cartodb/blob/79a6525bebfc866fad564852e7abd2277f4f85d9/lib/carto/ghost_tables_manager.rb#L23-L24

And also in the TIS:
https://github.com/CartoDB/node-cartodb-invalidation-service/blob/master/lib/backend/ghost_tables.js#L60

**Comment by gonzaloriestra** [comment]
Nice! But let's go step by step... These changes LGTM.

**Comment by gonzaloriestra** [comment]
And now let's continue the conversation in [the next PR](https://github.com/CartoDB/cartodb/pull/15276)...

**Comment by jgoizueta** [comment]
:+1: agreed, I closed the other PR with some comments, and I have created this ticket work on the single GT-job idea: #15277

**Pull request merged**